File: apps/api/src/pathfinder/ai/conversation/attachments.py
"""The files a message may carry to the model that reads it, and the caps on them."""

from __future__ import annotations

from collections.abc import Callable, Sequence
from dataclasses import dataclass

from pydantic import BaseModel, ConfigDict
from pydantic_ai import BinaryContent
from pydantic_ai.ui.vercel_ai.request_types import FileUIPart

from pathfinder.ai.models.catalog import ModelEntry, get_model_catalog
from pathfinder.platform.errors import (
    AttachmentNotReadableError,
    AttachmentTooLargeError,
)

_MIB = 1024 * 1024
MAX_FILE_BYTES = 10 * _MIB
MAX_MESSAGE_BYTES = 20 * _MIB
MAX_ATTACHMENTS = 6

IMAGE_MEDIA_TYPES = frozenset({"image/png", "image/jpeg", "image/webp", "image/gif"})
DOCUMENT_MEDIA_TYPES = frozenset({"application/pdf"})

_KINDS_READ = "PNG, JPEG, WebP and GIF images and PDF documents"


class ReadAttachment(BaseModel):
    """One attachment a turn takes: its name, its kind and its decoded size."""

    model_config = ConfigDict(frozen=True)

    filename: str
    media_type: str
    size: int


def _megabytes(size: int) -> str:
    return f"{size / _MIB:.1f} MB"


@dataclass(frozen=True)
class _Kind:
    noun: str
    media_types: frozenset[str]
    read_by: Callable[[ModelEntry], bool]


_KINDS = (
    _Kind("images", IMAGE_MEDIA_TYPES, lambda m: m.supports_images),
    _Kind("PDFs", DOCUMENT_MEDIA_TYPES, lambda m: m.supports_documents),
)


def _check_kind(part: FileUIPart, name: str, model: ModelEntry) -> None:
    kind = next((k for k in _KINDS if part.media_type in k.media_types), None)
    if kind is None:
        msg = f"{name} is {part.media_type}; PathFinder reads {_KINDS_READ}."
        raise AttachmentNotReadableError(msg)
    if kind.read_by(model):
        return
    readers = [m.name for m in get_model_catalog() if kind.read_by(m)]
    msg = (
        f"{model.name} does not read {kind.noun}; choose a model that does in "
        f"Settings: {', '.join(readers)}."
    )
    raise AttachmentNotReadableError(msg)


def _read_one(part: FileUIPart, model: ModelEntry) -> ReadAttachment:
    name = part.filename or "the attachment"
    _check_kind(part, name, model)
    if not part.url.startswith("data:"):
        msg = f"{name} is a link; an attachment is sent inline."
        raise AttachmentNotReadableError(msg)
    try:
        content = BinaryContent.from_data_uri(part.url)
    except ValueError as err:
        msg = f"{name} is not a readable data URL."
        raise AttachmentNotReadableError(msg) from err
    size = len(content.data)
    if size > MAX_FILE_BYTES:
        msg = (
            f"{name} is {_megabytes(size)}; one attachment can be at most "
            f"{MAX_FILE_BYTES // _MIB} MB."
        )
        raise AttachmentTooLargeError(msg)
    return ReadAttachment(filename=name, media_type=part.media_type, size=size)


def read_attachments(
    parts: Sequence[FileUIPart], model: ModelEntry
) -> list[ReadAttachment]:
    """The attachments of one message, each one readable by ``model``.

    :raises AttachmentTooLargeError: A file, or the message, is over its cap.
    :raises AttachmentNotReadableError: ``model`` cannot read a file's kind.
    """
    if len(parts) > MAX_ATTACHMENTS:
        msg = (
            f"One message can carry at most {MAX_ATTACHMENTS} attachments; "
            f"this one has {len(parts)}."
        )
        raise AttachmentTooLargeError(msg)
    read = [_read_one(part, model) for part in parts]
    total = sum(a.size for a in read)
    if total > MAX_MESSAGE_BYTES:
        msg = (
            f"These attachments come to {_megabytes(total)}; one message can carry "
            f"at most {MAX_MESSAGE_BYTES // _MIB} MB."
        )
        raise AttachmentTooLargeError(msg)
    return read


__all__ = [
    "DOCUMENT_MEDIA_TYPES",
    "IMAGE_MEDIA_TYPES",
    "MAX_ATTACHMENTS",
    "MAX_FILE_BYTES",
    "MAX_MESSAGE_BYTES",
    "ReadAttachment",
    "read_attachments",
]

File: apps/web/src/lib/models/attachments.ts
import type { ModelCatalogEntry } from "@pathfinder/shared";

// The api holds the same caps and kinds (ai/conversation/attachments.py) and
// refuses a message that breaks them; the composer says so before it is sent.

const MIB = 1024 * 1024;
export const MAX_ATTACHMENT_BYTES = 10 * MIB;
export const MAX_MESSAGE_ATTACHMENT_BYTES = 20 * MIB;
export const MAX_ATTACHMENTS = 6;

export type AttachmentKind = "gene-ids" | "image" | "pdf";

const GENE_ID_EXTENSIONS = [".csv", ".tsv", ".txt"];
const GENE_ID_TYPES = ["text/csv", "text/tab-separated-values", "text/plain"];
const IMAGE_TYPES = ["image/png", "image/jpeg", "image/webp", "image/gif"];
const PDF_TYPE = "application/pdf";

/** The role whose model reads the user's message, per assistant. */
const PROMPT_ROLES: Readonly<Record<string, string>> = {
  pathfinder: "lead",
  site_help: "site_help",
};

interface FileFacts {
  name: string;
  type: string;
}

export function attachmentKind(file: FileFacts): AttachmentKind | null {
  if (IMAGE_TYPES.includes(file.type)) return "image";
  if (file.type === PDF_TYPE) return "pdf";
  const lower = file.name.toLowerCase();
  if (GENE_ID_TYPES.includes(file.type)) return "gene-ids";
  if (GENE_ID_EXTENSIONS.some((ext) => lower.endsWith(ext))) return "gene-ids";
  return null;
}

function reads(model: ModelCatalogEntry, kind: "image" | "pdf"): boolean {
  return kind === "image"
    ? model.supportsImages === true
    : model.supportsDocuments === true;
}

/** The file chooser's accept list: gene-ID lists always, images and PDFs only
 *  when the reader reads them. */
export function acceptFor(model: ModelCatalogEntry | undefined): string {
  const accepted = [...GENE_ID_EXTENSIONS, ...GENE_ID_TYPES];
  if (model !== undefined && reads(model, "image")) accepted.push(...IMAGE_TYPES);
  if (model !== undefined && reads(model, "pdf")) accepted.push(PDF_TYPE);
  return accepted.join(",");
}

/** The files a model reads, as the model catalog lists it, or null for none. */
export function readsLabel(model: ModelCatalogEntry): string | null {
  const kinds = [
    ...(reads(model, "image") ? ["images"] : []),
    ...(reads(model, "pdf") ? ["PDFs"] : []),
  ];
  return kinds.length === 0 ? null : `reads ${kinds.join(" and ")}`;
}

/** The attach button's name: the kinds the reader takes. */
export function attachLabel(model: ModelCatalogEntry | undefined): string {
  const kinds = ["a gene-ID list"];
  if (model !== undefined && reads(model, "image")) kinds.push("an image");
  if (model !== undefined && reads(model, "pdf")) kinds.push("a PDF");
  const last = kinds.pop() ?? "";
  return kinds.length === 0
    ? `Attach ${last}`
    : `Attach ${kinds.join(", ")} or ${last}`;
}

/** Why the reader is offered no images or no PDFs, or null when it reads both. */
export function attachHint(model: ModelCatalogEntry | undefined): string | null {
  if (model === undefined) return null;
  const missing = [
    ...(reads(model, "image") ? [] : ["images"]),
    ...(reads(model, "pdf") ? [] : ["PDFs"]),
  ];
  if (missing.length === 0) return null;
  return (
    `${model.name} does not read ${missing.join(" or ")}; choose a model that ` +
    "does in Settings."
  );
}

/** What the composer says when the file chooser's accept list turned a file
 *  away: why this reader takes no image or PDF, else the kinds read here. */
export function notAcceptedSentence(model: ModelCatalogEntry | undefined): string {
  return (
    attachHint(model) ??
    "PathFinder reads gene-ID lists (.csv, .tsv, .txt), images (PNG, JPEG, " +
      "WebP, GIF) and PDFs."
  );
}

function megabytes(size: number): string {
  return `${(size / MIB).toFixed(1)} MB`;
}

/** Why one file cannot go with the message, or null when it can. The file
 *  chooser's accept list has already turned away a kind the reader does not
 *  read, so only the size is left to refuse. */
export function fileRefusal(file: { name: string; size: number }): string | null {
  if (file.size > MAX_ATTACHMENT_BYTES) {
    return (
      `${file.name} is ${megabytes(file.size)}; one attachment can be at most ` +
      `${MAX_ATTACHMENT_BYTES / MIB} MB.`
    );
  }
  return null;
}

/** Why the attachments of one message cannot go together, or null. */
export function messageRefusal(sizes: readonly number[]): string | null {
  if (sizes.length > MAX_ATTACHMENTS) {
    return (
      `One message can carry at most ${MAX_ATTACHMENTS} attachments; ` +
      `this one has ${sizes.length}.`
    );
  }
  const total = sizes.reduce((sum, size) => sum + size, 0);
  if (total > MAX_MESSAGE_ATTACHMENT_BYTES) {
    return (
      `These attachments come to ${megabytes(total)}; one message can carry ` +
      `at most ${MAX_MESSAGE_ATTACHMENT_BYTES / MIB} MB.`
    );
  }
  return null;
}

/** The model that reads the user's message: the pick for the reading role,
 *  else the deployment's default for it. */
export function readerModel(
  assistantId: string,
  picks: Readonly<Record<string, string>>,
  defaults: Readonly<Record<string, string>>,
  catalog: readonly ModelCatalogEntry[],
): ModelCatalogEntry | undefined {
  const role = PROMPT_ROLES[assistantId] ?? assistantId;
  const modelId = picks[role] ?? defaults[role];
  return catalog.find((m) => m.id === modelId);
}
